File: JS/LeetC/g75/ransomNote.py
from collections import Counter


def canConstruct(ransomNote, magazine):

    if len(ransomNote) > len(magazine):
        return False

    letters = Counter(magazine)

    for c in ransomNote:
        if not letters[c] or letters[c] <= 0:
            return False

    letters[c] = letters[c] - 1

    return True

File: JS/LeetC/g75/ransomNote.js
function canConstruct(ransomNote, magazine) {
  if (ransomNote.length > magazine.length) {
    return false;
  }
  const magazineFreq = frequencies(magazine);

  for (let c of ransomNote) {
    if (!magazineFreq[c] || magazineFreq[c] <= 0) {
      return false;
    }

    magazineFreq[c] = magazineFreq[c] - 1;
  }

  return true;
}

function frequencies(str) {
  const obj = {};
  for (let c of str) {
    obj[c] = obj[c] ? obj[c] + 1 : 1;
  }

  return obj;
}

main();

function main() {
  console.log(canConstruct('a', 'b'))
}
